File: Sorting/heapsort.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 100

void max_heapify(int a[], int n);
void build_max_heap(int a[]);
void swap(int a[], int x, int y);
void heapssort(int a[]);
void print_array(int a[]);
int heap_size;

void max_heapify(int a[], int n) {
  int l = 2*n+1;
  int r = 2*n+2;
  int largest = n;
  if (l <= heap_size-1 && a[l] > a[n])
    largest = l;
  if (r <= heap_size-1 && a[r] > a[largest])
    largest = r;
  if (largest != n) {
    swap(a, n, largest);
    max_heapify(a, largest);
  }
}

void build_max_heap(int a[]) {
  int i;
  for (i = heap_size/2 - 1; i >= 0; i--) {
    max_heapify(a,i);
  }
}

void swap(int a[], int x, int y) {
  int t;
  t = a[x]; a[x] = a[y]; a[y] = t;
}

void heapsort(int a[]) {
  heap_size = MAX;
  build_max_heap(a);
  while (heap_size > 1) {
    swap(a, 0, heap_size -1);
    --heap_size;
    max_heapify(a, 0);
  }
}

void print_array(int a[]) {
  int i;
  for (i = 0; i < MAX; i++) {
    printf("%i ", a[i]);
  }
  printf("\n");
}

int main(int argc, const char *argv[])
{
  int a[MAX];
  int i;
  for (i = 0; i < MAX; i++) {
    a[i] = rand()%100;
    printf("%i ", a[i]);
  }
  printf("\n");
  heapsort(a);
  print_array(a);
  return 0;
}

File: Sorting/HeapSort.java
public class HeapSort {
  private static int heapSize;
  public static void heapify(int[] arr, int n) {
    int left = 2*n + 1;
    int right = 2*n + 2;
    int largest = n;
    if (left <= heapSize && arr[left] > arr[largest])
      largest = left;
    if (right <= heapSize && arr[right] > arr[largest])
      largest = right;
    if (largest != n) {
      swap(arr, n, largest);
      heapify(arr, largest);
    }
  }

  public static void build_heap(int[] arr) {
    for (int i = (heapSize-1)/2; i >= 0; i--)
      heapify(arr, i);
  }

  public static void heapSort(int[] arr) {
    heapSize = arr.length-1;
    build_heap(arr);
    while (heapSize > 0) {
      swap(arr, 0, heapSize);
      --heapSize;
      heapify(arr, 0);
    }
  }

  public static void swap(int[] arr, int i, int k) {
    int t;
    t = arr[i]; arr[i] = arr[k]; arr[k] = t;
  }

  private static final int MAX = 10;

  public static void main (String[] args) {
    int[] array = new int[MAX];
    for (int x = 0; x != MAX; ++x) {
      array[x] = (int)(Math.random() * MAX);
    } 
    heapSort(array);
    for (int x : array) {
      System.out.print(x + " ");
    } 
    System.out.println();
  }
}
